Fix inefficient uses of Vector::fill()

cdumez opened the pull request:
#### 01b10b7f0e65f07e683f779755f4818487def6b8
<pre>
Fix inefficient uses of Vector::fill()
<a href="https://bugs.webkit.org/show_bug.cgi?id=252784">https://bugs.webkit.org/show_bug.cgi?id=252784</a>

Reviewed by Darin Adler.

Fix inefficient uses of Vector::fill(). Vector::fill() does a few things that
are unnecessary for a newly created Vector. It is more efficient to call the
`Vector(size, initialValue)` constructor.

* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
* Source/JavaScriptCore/b3/air/AirUseCounts.h:
(JSC::B3::Air::UseCounts::UseCounts):
* Source/JavaScriptCore/bytecode/Operands.h:
(JSC::Operands::Operands):
* Source/JavaScriptCore/dfg/DFGScoreBoard.h:
(JSC::DFG::ScoreBoard::ScoreBoard):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::JSWebAssemblyStruct):
* Source/WTF/wtf/FastBitVector.h:
(WTF::FastBitVector::FastBitVector):
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::FixedVector):
* Source/WTF/wtf/IndexMap.h:
(WTF::IndexMap::IndexMap):
* Source/WebCore/platform/audio/IIRFilter.cpp:
(WebCore::IIRFilter::IIRFilter):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::ReplicaState::cloneID const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::hasGlyphsForCharacterRange):
* Source/WebCore/rendering/RenderFrameSet.h:
(WebCore::FrameEdgeInfo::FrameEdgeInfo):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::RenderTable):

Canonical link: <a href="https://commits.webkit.org/260748@main">https://commits.webkit.org/260748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a2c941a89ad90e563140ea7efc2769c18126d11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109148 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18230 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/680 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118395 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113031 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9532 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101389 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114906 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14764 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97995 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42936 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96736 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84668 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/98157 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11010 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30992 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/98950 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9096 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11741 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7926 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31030 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50595 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106638 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7418 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13357 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26423 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -84,6 +84,12 @@ class FixedVector {
+    FixedVector(size_t size, const T& value)
+        : m_storage(size ? Storage::create(size).moveToUniquePtr() : nullptr)
+    {
+        fill(value);
```
I'm a little puzzled by this. The change log implies we are avoiding using `fill` so we can use more efficient constructors. But this constructor simply uses `fill`. Perhaps the main value in this change is avoiding initializing twice? Or are the constructors more efficient?

Or maybe the main problem is the caller who use `fill` to **resize** a vector. That’s definitely less efficient than sizing it correctly in the first place.

darinadler replied:
Are there **any** good uses of `fill`?

cdumez replied:
The perf benefits are for Vector only. The reason I added this constructor to FixedVector and BitVector is purely because one of the classes in JSC takes the vector type as template parameter and so I needed consistency in those vectors apis for that class to benefit from the more efficient Vector constructor. I think the consistency between those vector types is a plus too.

cdumez replied:
And yes, they are plenty of remaining good uses of fill(), on vectors that were not just created. Often in functions that reset or update things.

cdumez commented on Source/JavaScriptCore/bytecode/Operands.h:
```diff
@@ -158,20 +158,18 @@ class Operands {
 
     explicit Operands(size_t numArguments, size_t numLocals, size_t numTmps, const T& initialValue)
-        : m_values(numArguments + numLocals + numTmps)
+        : m_values(numArguments + numLocals + numTmps, initialValue)
```
This class here is the one that takes the vector type as template parameter. I believe it can get constructed with a Vector since I made this change but I will double check to make sure the changes to BitVector and FixedVector are really needed.

cdumez committed:
Fix inefficient uses of Vector::fill()
https://bugs.webkit.org/show_bug.cgi?id=252784

Reviewed by Darin Adler.

Fix inefficient uses of Vector::fill(). Vector::fill() does a few things that
are unnecessary for a newly created Vector. It is more efficient to call the
`Vector(size, initialValue)` constructor.

* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
* Source/JavaScriptCore/b3/air/AirUseCounts.h:
(JSC::B3::Air::UseCounts::UseCounts):
* Source/JavaScriptCore/bytecode/Operands.h:
(JSC::Operands::Operands):
* Source/JavaScriptCore/dfg/DFGScoreBoard.h:
(JSC::DFG::ScoreBoard::ScoreBoard):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::JSWebAssemblyStruct):
* Source/WTF/wtf/FastBitVector.h:
(WTF::FastBitVector::FastBitVector):
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::FixedVector):
* Source/WTF/wtf/IndexMap.h:
(WTF::IndexMap::IndexMap):
* Source/WebCore/platform/audio/IIRFilter.cpp:
(WebCore::IIRFilter::IIRFilter):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::ReplicaState::cloneID const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::hasGlyphsForCharacterRange):
* Source/WebCore/rendering/RenderFrameSet.h:
(WebCore::FrameEdgeInfo::FrameEdgeInfo):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::RenderTable):

Canonical link: https://commits.webkit.org/260748@main
